zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/websocket/ZeppelinClient.java

```diff
@@ -0,0 +1,264 @@
+    Session noteSession = getZeppelinConnection(noteId);
+    if (!isSessionOpen(noteSession)) {
+      LOG.error("Cannot open websocket connection to Zeppelin note {}", noteId);
+      return;
+    }
+    noteSession.getRemote().sendStringByFuture(serialize(msg));
+  }
+
+  private boolean isSessionOpen(Session session) {
+    return (session != null) && (session.isOpen());
+  }
+  
+  /* per notebook based ws connection, returns null if can't connect */
+  public Session getZeppelinConnection(String noteId) {
+    if (StringUtils.isBlank(noteId)) {
+      LOG.warn("Cannot return websocket connection for blank noteId");
+      return null;
+    }
+
+    if (zeppelinConnectionMap.containsKey(noteId)) {
+      LOG.info("Connection for {} exists in map", noteId);
+      return getNoteSession(noteId);
+    }
+    //TODO(khalid): clean log later
+    LOG.info("Creating Zeppelin websocket connection {} {}", zeppelinWebsocketUrl, noteId);
+    return openNoteSession(noteId);
+  }
+
+  private Message zeppelinGetNoteMsg(String noteId) {
+    Message getNoteMsg = new Message(Message.OP.GET_NOTE);
+    HashMap<String, Object> data = new HashMap<String, Object>();
+    data.put("id", noteId);
+    getNoteMsg.data = data;
+    return getNoteMsg;
+  }
+  
+  private Session getNoteSession(String noteId) {
+    Session session = zeppelinConnectionMap.get(noteId);
+    if (session == null || !session.isOpen()) {
+      LOG.info("Not connection to {}", noteId);
+      zeppelinConnectionMap.remove(noteId);
+      session = openNoteSession(noteId);
+    }
+    return session;
+  }
+  
+  private Session openNoteSession(String noteId) {
+    ClientUpgradeRequest request = new ClientUpgradeRequest();
+    ZeppelinWebsocket socket = new ZeppelinWebsocket(noteId);
+    Future<Session> future = null;
+    Session session = null;
+    try {
+      future = wsClient.connect(socket, zeppelinWebsocketUrl, request);
+      session = future.get();
+    } catch (IOException | InterruptedException | ExecutionException e) {
+      LOG.error("Couldn't establish websocket connection to Zeppelin ", e);
+      return session;
+    }
+
+    if (zeppelinConnectionMap.containsKey(noteId)) {
+      session.close();
+      session = zeppelinConnectionMap.get(noteId);
+    } else {
+      String getNote = serialize(zeppelinGetNoteMsg(noteId));
+      // TODO(khalid): may need to check return whether successful
+      session.getRemote().sendStringByFuture(getNote);
+      zeppelinConnectionMap.put(noteId, session);
+    }
+    return session;
+  }
+
+  public void handleMsgFromZeppelin(String message, String noteId) {
+    Map<String, String> meta = new HashMap<String, String>();
+    meta.put("token", zeppelinhubToken);
+    meta.put("noteId", noteId);
+    Message zeppelinMsg = deserialize(message);
+    if (zeppelinMsg == null) {
+      return;
+    }
+    ZeppelinhubMessage hubMsg = ZeppelinhubMessage.newMessage(zeppelinMsg, meta);
+    Client client = Client.getInstance();
```

Comment:
Instead of introducing static dependency here, would not it rather make sense to have `Client` instance injected as a regular dependency for the `ZeppelinClient` thought the constructor? Does it have downsides?

The benefits of explicit dependencies are big - testability for one, and it's easier to discover and follow the code with explicit dependencies.